subsys/bluetooth/services/ias.c

```diff
@@ -0,0 +1,143 @@
+/** @file
+ *  @brief Immediate Alert Service implementation
+ */
+
+/*
+ * Copyright (c) 2022 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <init.h>
+#include <stdint.h>
+#include <zephyr.h>
+#include <net/buf.h>
+#include <logging/log.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/services/ias.h>
+
+#define LOG_LEVEL CONFIG_BT_IAS_LOG_LEVEL
+LOG_MODULE_REGISTER(ias);
+
+#define BT_IAS_ALERT_MAX_LEN 1
+
+#if defined(CONFIG_BT_IAS_DEFAULT_PERM_W_AUTHEN)
+#define IAS_GATT_PERM_DEFAULT BT_GATT_PERM_WRITE_AUTHEN
+#elif defined(CONFIG_BT_IAS_DEFAULT_PERM_W_ENCRYPT)
+#define IAS_GATT_PERM_DEFAULT CONFIG_BT_IAS_DEFAULT_PERM_W_ENCRYPT
+#else
+#define IAS_GATT_PERM_DEFAULT BT_GATT_PERM_WRITE
+#endif
+
+typedef enum __packed {
+	BT_IAS_ALERT_LVL_NO_ALERT,
+	BT_IAS_ALERT_LVL_MILD_ALERT,
+	BT_IAS_ALERT_LVL_HIGH_ALERT,
+} bt_ias_alert_lvl_t;
+
+struct alerting_device {
+	bt_ias_alert_lvl_t alert_level;
+};
+
+static struct alerting_device devices[CONFIG_BT_MAX_CONN];
+static bt_ias_alert_lvl_t curr_lvl;
+
+static void set_alert_level()
+{
+	bt_ias_alert_lvl_t alert_level;
+
+	alert_level = devices[0].alert_level;
+	for (int i = 1; i < CONFIG_BT_MAX_CONN; i++) {
+		if (alert_level < devices[i].alert_level) {
+			alert_level = devices[i].alert_level;
+		}
+	}
+
+	if (curr_lvl == alert_level) {
+		return;
+	}
+
+	if (alert_level == BT_IAS_ALERT_LVL_HIGH_ALERT) {
+		STRUCT_SECTION_FOREACH(bt_ias_cb, cb) {
+			if (cb->high_alert) {
+				cb->high_alert();
+			}
+		}
+		LOG_DBG("High alert");
+	} else if (alert_level == BT_IAS_ALERT_LVL_MILD_ALERT) {
+		STRUCT_SECTION_FOREACH(bt_ias_cb, cb) {
+			if (cb->mild_alert) {
+				cb->mild_alert();
+			}
+		}
+		LOG_DBG("Mild alert");
+	} else {
+		STRUCT_SECTION_FOREACH(bt_ias_cb, cb) {
+			if (cb->no_alert) {
+				cb->no_alert();
+			}
+		}
+		LOG_DBG("No alert");
+	}
+	curr_lvl = alert_level;
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	devices[bt_conn_index(conn)].alert_level = BT_IAS_ALERT_LVL_NO_ALERT;
+	set_alert_level();
+}
+
+int local_alert_stop()
```

Comment:
Given that we now have multiple alert levels (one for each connection) perhaps this function should also take a `struct bt_conn` parameter. 

We could then also allow settings it to NULL to setting the alert level for all connections.

Reply:
I'm having some doubts regarding your ideas for this service. Yes, the IAS specification is pretty lacking in terms of handling multiple devices and the outcomes of various alerting possibilities. Nevertheless, I think that we should not make this service a "swiss-army knife" contraption. I believe we shouldn't over-complicate this by passing information to the application about specific devices and their alert levels, I think this should be done within the service and it should follow the rule that the current highest alert-level is the one that is paramount.

Reply:
![Screenshot from 2022-03-25 15-33-47](https://user-images.githubusercontent.com/15470986/160141629-ec8b9736-1b66-45a6-9bca-61738bacd906.png)